# Lesson 36: Data Cleaning


|Particulars|Description|
|-|-|
|**Topics Covered**|Replacing Commas|
||Garbage Value Inspection|
||Garbage Value Replacement|
|||
|**Lesson Description**|In this class, a student will learn to clean a DataFrame by treating garbage value.|
|||
|**Lesson Duration**|45 minutes|
|||
|**Learning Outcomes**|Replace the values containing comma with periods.|
||Replace the `-200` value with the median values for each column.|
||Slice a DataFrame into two DataFrames.|
|||


---

### Teacher-Student Tasks

In this class, we will clean the DataFrame by replacing commas with periods (or dots) and by replacing the `-200` garbage value with the median value of the respective columns.

The data cleaning exercises depend on the kind of dataset you get. However, the functions applied or the operations followed to clean a DataFrame more or less remain the same.

Let's begin the class by running the code which we have already covered in the previous classes and continue the class from the **Task 1: Replacing Commas** section.

---

### Recap

Run the code cell below to load the dataset and apply the operations on the DataFrame that were covered in the previous classes.

In [ ]:
# Run the code cell.
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

# Loading the dataset.
csv_file = 'https://student-datasets-bucket.s3.ap-south-1.amazonaws.com/whitehat-ds-datasets/air-quality/AirQualityUCI.csv'
df = pd.read_csv(csv_file, sep=';')

# Dropping the 'Unnamed: 15' & 'Unnamed: 16' columns.
df = df.drop(columns=['Unnamed: 15', 'Unnamed: 16'], axis=1)

# Dropping the null values.
df = df.dropna()

# Creating a Pandas series containing 'datetime' objects.
dt_series = pd.Series(data = [item.split("/")[2] + "-" + item.split("/")[1] + "-" + item.split("/")[0] for item in df['Date']], index=df.index) + ' ' + pd.Series(data=[str(item).replace(".", ":") for item in df['Time']], index=df.index)
dt_series = pd.to_datetime(dt_series)

# Remove the Date & Time columns from the DataFrame and insert the 'dt_series' in it.
df = df.drop(columns=['Date', 'Time'], axis=1)
df.insert(loc=0, column='DateTime', value=dt_series)

# Get the Pandas series containing the year values as integers.
year_series = dt_series.dt.year

# Get the Pandas series containing the month values as integers.
month_series = dt_series.dt.month

# Get the Pandas series containing the day values as integers.
day_series = dt_series.dt.day

# Get the Pandas series containing the days of a week, i.e., Monday, Tuesday, Wednesday etc.
day_name_series = dt_series.dt.day_name()

# Add the 'Year', 'Month', 'Day' and 'Day Name' columns to the DataFrame.
df['Year'] = year_series
df['Month'] = month_series
df['Day'] = day_series
df['Day Name'] = day_name_series

# Sort the DataFrame by the 'DateTime' values in the ascending order. Also, display the first 10 rows of the DataFrame.
df = df.sort_values(by='DateTime')

---

#### Task 1: Replacing Commas

The  values in the `CO(GT), C6H6(GT), T,	RH` and	`AH` columns contain the commas in them. Let's replace the commas with periods (or dots).

In [ ]:
# S1.1: Display the first five rows of the DataFrame.
df.head()

,DateTime,CO(GT),PT08.S1(CO),NMHC(GT),C6H6(GT),PT08.S2(NMHC),NOx(GT),PT08.S3(NOx),NO2(GT),PT08.S4(NO2),PT08.S5(O3),T,RH,AH,Year,Month,Day,Day Name
0,2004-03-10 18:00:00,"2,6",1360.0,150.0,"11,9",1046.0,166.0,1056.0,113.0,1692.0,1268.0,"13,6","48,9","0,7578",2004,3,10,Wednesday
1,2004-03-10 19:00:00,2,1292.0,112.0,"9,4",955.0,103.0,1174.0,92.0,1559.0,972.0,"13,3","47,7","0,7255",2004,3,10,Wednesday
2,2004-03-10 20:00:00,"2,2",1402.0,88.0,"9,0",939.0,131.0,1140.0,114.0,1555.0,1074.0,"11,9","54,0","0,7502",2004,3,10,Wednesday
3,2004-03-10 21:00:00,"2,2",1376.0,80.0,"9,2",948.0,172.0,1092.0,122.0,1584.0,1203.0,"11,0","60,0","0,7867",2004,3,10,Wednesday
4,2004-03-10 22:00:00,"1,6",1272.0,51.0,"6,5",836.0,131.0,1205.0,116.0,1490.0,1110.0,"11,2","59,6","0,7888",2004,3,10,Wednesday


Create a function which takes a Pandas series containing comma separated decimal values as an input and returns a new Pandas series containing period separated decimal values as an output.

In [ ]:
# S1.2: Create a function to replace the commas with periods in a Pandas series.

def comma_to_period(ser):
  p = pd.Series([float(str(i).replace(',','.')) for i in ser],index = df.index)
  return p


In the above code,

- we iterated through each item of a `series`,

- converted each item to a string value using the `str()` function,

- replaced the comma with the period using the `replace()` function

- converted the replaced value to a floating-point number using the `float()` function,

- added each item to a Python list, and

- converted the list to a Pandas series using the `pd.Series()` function.

We have ensured that the indices of the new Pandas series are the same as the indices of the DataFrame by passing `index = df.index` parameter to the `pd.Series()` function where `df` is the DataFrame.

Now, let's test the `comma_to_period()` function by correcting the values of the `CO(GT)` column but storing the output in a new variable so that the original values of the column are unaffected before we have verified the correctness of the function.

In [ ]:
# S1.3: Test the 'comma_to_period()' function by correcting the values of 'CO(GT)' column but storing the output in a new variable.
print(comma_to_period(df['CO(GT)']))

0       2.6
1       2.0
2       2.2
3       2.2
4       1.6
       ... 
9352    3.1
9353    2.4
9354    2.4
9355    2.1
9356    2.2
Length: 9357, dtype: float64


As you can see, all the values of the `CO(GT)` column are now proper float-point numbers.

Let's apply the `comma_to_period()` function on the `'CO(GT), C6H6(GT), T, RH` and `AH` columns with the help of the following steps:

1. Create a list of all the columns whose values we need to correct.

2. Iterate through each value of the list.

3. Apply the `comma_to_period()` function on each column of the `df` DataFrame.

4. Replace each column with their corresponding new series to correct the required values.


In [ ]:
# S1.4: Apply the 'comma_to_period()' function on the ''CO(GT)', 'C6H6(GT)', 'T', 'RH' and 'AH' columns.
l = ['CO(GT)', 'C6H6(GT)', 'T', 'RH','AH']
for i in l:
  df[i] = comma_to_period(df[i])

Let's print the first five rows of the DataFrame to see the effect of the `comma_to_period()` function:


In [ ]:
# S1.5: Display the first five rows of the DataFrame to see the effect of the 'comma_to_period()' function.
df.head()

,DateTime,CO(GT),PT08.S1(CO),NMHC(GT),C6H6(GT),PT08.S2(NMHC),NOx(GT),PT08.S3(NOx),NO2(GT),PT08.S4(NO2),PT08.S5(O3),T,RH,AH,Year,Month,Day,Day Name
0,2004-03-10 18:00:00,2.6,1360.0,150.0,11.9,1046.0,166.0,1056.0,113.0,1692.0,1268.0,13.6,48.9,0.7578,2004,3,10,Wednesday
1,2004-03-10 19:00:00,2.0,1292.0,112.0,9.4,955.0,103.0,1174.0,92.0,1559.0,972.0,13.3,47.7,0.7255,2004,3,10,Wednesday
2,2004-03-10 20:00:00,2.2,1402.0,88.0,9.0,939.0,131.0,1140.0,114.0,1555.0,1074.0,11.9,54.0,0.7502,2004,3,10,Wednesday
3,2004-03-10 21:00:00,2.2,1376.0,80.0,9.2,948.0,172.0,1092.0,122.0,1584.0,1203.0,11.0,60.0,0.7867,2004,3,10,Wednesday
4,2004-03-10 22:00:00,1.6,1272.0,51.0,6.5,836.0,131.0,1205.0,116.0,1490.0,1110.0,11.2,59.6,0.7888,2004,3,10,Wednesday


Let's also print the information on the `df` DataFrame. Except for the first column and the last column, all other columns must have the numeric (`float` or `int`) data-type values.

In [ ]:
# S1.6: Print the first information of the 'df' DataFrame.
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 9357 entries, 0 to 9356
Data columns (total 18 columns):
 #   Column         Non-Null Count  Dtype         
---  ------         --------------  -----         
 0   DateTime       9357 non-null   datetime64[ns]
 1   CO(GT)         9357 non-null   float64       
 2   PT08.S1(CO)    9357 non-null   float64       
 3   NMHC(GT)       9357 non-null   float64       
 4   C6H6(GT)       9357 non-null   float64       
 5   PT08.S2(NMHC)  9357 non-null   float64       
 6   NOx(GT)        9357 non-null   float64       
 7   PT08.S3(NOx)   9357 non-null   float64       
 8   NO2(GT)        9357 non-null   float64       
 9   PT08.S4(NO2)   9357 non-null   float64       
 10  PT08.S5(O3)    9357 non-null   float64       
 11  T              9357 non-null   float64       
 12  RH             9357 non-null   float64       
 13  AH             9357 non-null   float64       
 14  Year           9357 non-null   int64         
 15  Month          9357 n

Hence, we have successfully replaced the commas with periods in the following columns:

- `CO(GT)`

- `C6H6(GT)`

-  `T`

- `RH`

- `AH`



---


#### Task 2: Garbage Value Inspection

The columns in the `df` DataFrame also contain `-200` value. It is a garbage value or just a random number to represent the further missing (or null) values in the DataFrame. Let's replace it with the most appropriate values for each column.

In [ ]:
# S2.1: Get the descriptive statistics for all the numeric data-type columns.
df.describe()

,CO(GT),PT08.S1(CO),NMHC(GT),C6H6(GT),PT08.S2(NMHC),NOx(GT),PT08.S3(NOx),NO2(GT),PT08.S4(NO2),PT08.S5(O3),T,RH,AH,Year,Month,Day
count,9357.000000,9357.000000,9357.000000,9357.000000,9357.000000,9357.000000,9357.000000,9357.000000,9357.000000,9357.000000,9357.000000,9357.000000,9357.000000,9357.000000,9357.000000,9357.000000
mean,-34.207524,1048.990061,-159.090093,1.865683,894.595276,168.616971,794.990168,58.148873,1391.479641,975.072032,9.778305,39.485380,-6.837604,2004.240141,6.310356,15.876884
std,77.657170,329.832710,139.789093,41.380206,342.333252,257.433866,321.993552,126.940455,467.210125,456.938184,43.203623,51.216145,38.976670,0.427192,3.438160,8.808653
min,-200.000000,-200.000000,-200.000000,-200.000000,-200.000000,-200.000000,-200.000000,-200.000000,-200.000000,-200.000000,-200.000000,-200.000000,-200.000000,2004.000000,1.000000,1.000000
25%,0.600000,921.000000,-200.000000,4.000000,711.000000,50.000000,637.000000,53.000000,1185.000000,700.000000,10.900000,34.100000,0.692300,2004.000000,3.000000,8.000000
50%,1.500000,1053.000000,-200.000000,7.900000,895.000000,141.000000,794.000000,96.000000,1446.000000,942.000000,17.200000,48.600000,0.976800,2004.000000,6.000000,16.000000
75%,2.600000,1221.000000,-200.000000,13.600000,1105.000000,284.000000,960.000000,133.000000,1662.000000,1255.000000,24.100000,61.900000,1.296200,2004.000000,9.000000,23.000000
max,11.900000,2040.000000,1189.000000,63.700000,2214.000000,1479.000000,2683.000000,340.000000,2775.000000,2523.000000,44.600000,88.700000,2.231000,2005.000000,12.000000,31.000000


Except for the `Year, Month,` and `Day` columns all the other columns contain the `-200` as its minimum value. Because of thi, their mean values are also affected. The median value never gets affected due to some garbage value or very high and very low values. So, we can replace `-200` with the median value for each column.

But first, let's find out how many rows contain `-200` in each column except for the `DateTime, Year, Month,` and `Day` columns with the help of the following steps:

1. Iterate through each column having indices between `1` and `-4` (excluding `-4`) in the `df` DataFrame.

2. Retrieve the rows containing `-200` value using the `df[df[col] == -200]` operation.

3. Calculate the number of such rows using the `shape[0]` attribute.

4. Add the column name and the corresponding counts of `-200` in each column as a tuple in a list using the list comprehension method.



In [ ]:
# S2.2: How many rows contain -200 in each column except for the 'DateTime', 'Year', 'Month' and 'Day' columns?
l1 = [(i,df[df[i] == -200].shape[0]) for i in df.columns[1:-4]]
l1

[('CO(GT)', 1683),
 ('PT08.S1(CO)', 366),
 ('NMHC(GT)', 8443),
 ('C6H6(GT)', 366),
 ('PT08.S2(NMHC)', 366),
 ('NOx(GT)', 1639),
 ('PT08.S3(NOx)', 366),
 ('NO2(GT)', 1642),
 ('PT08.S4(NO2)', 366),
 ('PT08.S5(O3)', 366),
 ('T', 366),
 ('RH', 366),
 ('AH', 366)]

In [ ]:
l3 = [(i,df[df[i] == -200]) for i in df.columns[1:-4]]
l3

[('PT08.S1(CO)',
                  DateTime  PT08.S1(CO)  C6H6(GT)  ...  Month  Day  Day Name
  524  2004-04-01 14:00:00       -200.0    -200.0  ...      4    1  Thursday
  525  2004-04-01 15:00:00       -200.0    -200.0  ...      4    1  Thursday
  526  2004-04-01 16:00:00       -200.0    -200.0  ...      4    1  Thursday
  701  2004-04-08 23:00:00       -200.0    -200.0  ...      4    8  Thursday
  702  2004-04-09 00:00:00       -200.0    -200.0  ...      4    9    Friday
  ...                  ...          ...       ...  ...    ...  ...       ...
  8111 2005-02-11 17:00:00       -200.0    -200.0  ...      2   11    Friday
  8112 2005-02-11 18:00:00       -200.0    -200.0  ...      2   11    Friday
  8113 2005-02-11 19:00:00       -200.0    -200.0  ...      2   11    Friday
  8114 2005-02-11 20:00:00       -200.0    -200.0  ...      2   11    Friday
  8777 2005-03-11 11:00:00       -200.0    -200.0  ...      3   11    Friday
  
  [366 rows x 14 columns]),
 ('C6H6(GT)',
              

As you can see we have got the exact number of rows containing `-200` in each column except for the `DateTime`, `Year`, `Month`, and `Day` columns.

Now, let's also find out the percentage of rows containing `-200` in each column except for the `DateTime, Year, Month,` and `Day` columns:

In [ ]:
# S2.3: Find out the percentage of rows containing -200 in each column except for the 'DateTime', 'Year', 'Month' and 'Day' columns.
l4 = [(i,round((df[df[i]==-200].shape[0]/df.shape[0])*100)) for i in df.columns[1:-4]]
l4

[('CO(GT)', 18),
 ('PT08.S1(CO)', 4),
 ('NMHC(GT)', 90),
 ('C6H6(GT)', 4),
 ('PT08.S2(NMHC)', 4),
 ('NOx(GT)', 18),
 ('PT08.S3(NOx)', 4),
 ('NO2(GT)', 18),
 ('PT08.S4(NO2)', 4),
 ('PT08.S5(O3)', 4),
 ('T', 4),
 ('RH', 4),
 ('AH', 4)]

Using the same process as above, we calculate the percentage of rows containing `-200` in each column except for the `DateTime, Year, Month` and `Day` columns.

Now, let's remove all the columns from the `df` DataFrame containing more than 10% garbage value.

In [ ]:
# S2.4: Remove all the columns from the 'df' DataFrame containing more than 10% garbage value.
df = df.drop(columns=['CO(GT)','NMHC(GT)','NOx(GT)','NO2(GT)'])

Again, calculate the percentage of rows containing `-200` in each column except for the `DateTime, Year, Month` and `Day` columns.

In [ ]:
# S2.5: Again, calculate the percentage of rows containing '-200' in each column except for the 'DateTime', 'Year', 'Month' and 'Day' columns.
l4 = [(i,round((df[df[i]==-200].shape[0]/df.shape[0])*100)) for i in df.columns[1:-4]]
l4

[('PT08.S1(CO)', 4),
 ('C6H6(GT)', 4),
 ('PT08.S2(NMHC)', 4),
 ('PT08.S3(NOx)', 4),
 ('PT08.S4(NO2)', 4),
 ('PT08.S5(O3)', 4),
 ('T', 4),
 ('RH', 4),
 ('AH', 4)]

Let's replace the `-200` value with the median values in all the columns except for the `DateTime, Year, Month` and `Day` columns.

Before that we should split the entire DataFrame in two different DataFrames because it contains data for two different years, i.e., 2004 and 2005. Then we should calculate the median values for each column for 2004 and 2005 separately. However, we can also first find out whether the median values are actually different for the two years. If they are not, then we don't need to split the DataFrame into two DataFrames for 2004 and 2005 data points.

In [ ]:
# S2.6: Calculate the median values for the columns having indices between 1 and -4 (excluding -4) for the year 2004.
df.loc[df['Year'] == 2004,df.columns[1:-4]].describe()

,PT08.S1(CO),C6H6(GT),PT08.S2(NMHC),PT08.S3(NOx),PT08.S4(NO2),PT08.S5(O3),T,RH,AH
count,7110.000000,7110.000000,7110.00000,7110.000000,7110.000000,7110.000000,7110.000000,7110.000000,7110.000000
mean,1055.135443,3.876779,923.60661,822.454571,1495.659353,973.985513,13.779001,40.292053,-5.310343
std,314.719517,37.859514,332.55071,315.705690,433.109562,434.094624,39.705743,46.983126,35.441047
min,-200.000000,-200.000000,-200.00000,-200.000000,-200.000000,-200.000000,-200.000000,-200.000000,-200.000000
25%,919.000000,4.600000,741.00000,666.000000,1331.250000,712.000000,14.300000,33.000000,0.855925
50%,1052.000000,8.600000,923.00000,815.000000,1528.000000,933.000000,20.150000,47.700000,1.083550
75%,1221.000000,14.200000,1124.00000,982.000000,1726.000000,1233.750000,25.800000,61.000000,1.404175
max,2040.000000,63.700000,2214.00000,2683.000000,2775.000000,2523.000000,44.600000,88.700000,2.231000


In [ ]:
# S2.7: Calculate the median values for the columns having indices between 1 and -4 (excluding -4) for the year 2005.
df.loc[df['Year'] == 2005,df.columns[1:-4]].describe()

,PT08.S1(CO),C6H6(GT),PT08.S2(NMHC),PT08.S3(NOx),PT08.S4(NO2),PT08.S5(O3),T,RH,AH
count,2247.000000,2247.000000,2247.000000,2247.000000,2247.000000,2247.000000,2247.000000,2247.000000,2247.000000
mean,1029.544726,-4.497864,802.797063,708.086782,1061.832221,978.510013,-2.880774,36.932888,-11.670190
std,373.059005,50.426047,356.435107,326.303957,415.312391,522.776411,50.737736,62.700754,48.186199
min,-200.000000,-200.000000,-200.000000,-200.000000,-200.000000,-200.000000,-200.000000,-200.000000,-200.000000
25%,925.500000,2.900000,643.500000,553.500000,906.000000,641.000000,4.700000,37.200000,0.413350
50%,1055.000000,5.800000,805.000000,721.000000,1084.000000,973.000000,8.600000,50.900000,0.596400
75%,1222.000000,11.200000,1021.500000,893.500000,1297.000000,1344.500000,13.400000,64.750000,0.804950
max,1846.000000,43.000000,1831.000000,1881.000000,2147.000000,2494.000000,30.000000,86.600000,1.393000


As we can see, the median values for 2004 and 2005 are different. Hence, we should split the `df` DataFrame into two different DataFrames. One for 2004 data points and another for 2005 data points.

---

#### Task 3: Garbage Value Replacement

Before we separate the DataFrame, let's calculate the count of 2004 and 2005 data points.

In [ ]:
# S3.1: Count the number of rows containing '2004' and '2005' year values.
df['Year'].value_counts()

2004    7110
2005    2247
Name: Year, dtype: int64

The DataFrame for 2004 year should have 7110 rows which means the DataFrame for 2005 year should have 2247 rows.

In [ ]:
# S3.2: Create a new DataFrame containing records for the year 2004. Also, display the first five rows.
df_2004 = df[df['Year']==2004]
df_2004

,DateTime,PT08.S1(CO),C6H6(GT),PT08.S2(NMHC),PT08.S3(NOx),PT08.S4(NO2),PT08.S5(O3),T,RH,AH,Year,Month,Day,Day Name
0,2004-03-10 18:00:00,1360.0,11.9,1046.0,1056.0,1692.0,1268.0,13.6,48.9,0.7578,2004,3,10,Wednesday
1,2004-03-10 19:00:00,1292.0,9.4,955.0,1174.0,1559.0,972.0,13.3,47.7,0.7255,2004,3,10,Wednesday
2,2004-03-10 20:00:00,1402.0,9.0,939.0,1140.0,1555.0,1074.0,11.9,54.0,0.7502,2004,3,10,Wednesday
3,2004-03-10 21:00:00,1376.0,9.2,948.0,1092.0,1584.0,1203.0,11.0,60.0,0.7867,2004,3,10,Wednesday
4,2004-03-10 22:00:00,1272.0,6.5,836.0,1205.0,1490.0,1110.0,11.2,59.6,0.7888,2004,3,10,Wednesday
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
7105,2004-12-31 19:00:00,930.0,5.3,781.0,933.0,906.0,866.0,9.3,33.6,0.3958,2004,12,31,Friday
7106,2004-12-31 20:00:00,962.0,5.3,780.0,919.0,909.0,884.0,9.6,32.3,0.3871,2004,12,31,Friday
7107,2004-12-31 21:00:00,974.0,5.5,790.0,915.0,895.0,951.0,9.1,32.5,0.3766,2004,12,31,Friday
7108,2004-12-31 22:00:00,1055.0,5.6,791.0,845.0,936.0,1195.0,9.1,37.2,0.4310,2004,12,31,Friday


In [ ]:
# S3.3 Calculate the number of rows and columns in the DataFrame for the 2004 year records.
df_2004.shape

(7110, 14)

In [ ]:
# S3.4: Create a new DataFrame containing records for the year 2005. Also, display the first five rows.
df_2005 = df[df['Year'] == 2005]
df_2005

,DateTime,PT08.S1(CO),C6H6(GT),PT08.S2(NMHC),PT08.S3(NOx),PT08.S4(NO2),PT08.S5(O3),T,RH,AH,Year,Month,Day,Day Name
7110,2005-01-01 00:00:00,1046.0,4.2,724.0,848.0,898.0,1201.0,8.2,40.1,0.4375,2005,1,1,Saturday
7111,2005-01-01 01:00:00,1275.0,8.8,930.0,649.0,1024.0,1617.0,5.3,50.7,0.4564,2005,1,1,Saturday
7112,2005-01-01 02:00:00,1173.0,7.5,878.0,738.0,1002.0,1355.0,5.9,50.0,0.4689,2005,1,1,Saturday
7113,2005-01-01 03:00:00,1163.0,7.6,881.0,748.0,1001.0,1296.0,4.9,53.9,0.4693,2005,1,1,Saturday
7114,2005-01-01 04:00:00,1054.0,5.6,791.0,830.0,967.0,1131.0,4.3,55.3,0.4650,2005,1,1,Saturday
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
9352,2005-04-04 10:00:00,1314.0,13.5,1101.0,539.0,1374.0,1729.0,21.9,29.3,0.7568,2005,4,4,Monday
9353,2005-04-04 11:00:00,1163.0,11.4,1027.0,604.0,1264.0,1269.0,24.3,23.7,0.7119,2005,4,4,Monday
9354,2005-04-04 12:00:00,1142.0,12.4,1063.0,603.0,1241.0,1092.0,26.9,18.3,0.6406,2005,4,4,Monday
9355,2005-04-04 13:00:00,1003.0,9.5,961.0,702.0,1041.0,770.0,28.3,13.5,0.5139,2005,4,4,Monday


In [ ]:
# S3.5 Calculate the number of rows and columns in the DataFrame for the 2004 year records.
df_2005.shape

(2247, 14)

Now, let's replace the `-200` value with the median values for each column having indices between 1 and -4 (excluding -4) for both the 2004 and 2005 year DataFrames.

In [ ]:
# S3.6: Replace the -200 value with the median values for each column having indices between 1 and -4 (excluding -4) for the 2004 year DataFrame.
import warnings
warnings.filterwarnings('ignore')

for i in df_2004.columns[1:-4]:
  med = df_2004.loc[df[i] != -200,i].median()
  df_2004[i] = df_2004.replace(to_replace=-200,value=med)



**Note**: The below code will remove all the warnings that we may get.

```
import warnings
warnings.filterwarnings('ignore')
```



In [ ]:
# S3.7: Repeat the same exercise for the 2005 year DataFrame.
for i in df_2005.columns[1:-4]:
  med = df_2005.loc[df_2005[i] == -200,i].median()
  df_2005[i] = df_2005.replace(to_replace=-200,value=med)

Now, let's again, calculate the number of rows containing '-200' in each column except for the `DateTime, Year, Month` and `Day` columns in both the DataFrames.

In [ ]:
# S3.8: Compute the number of rows containing '-200' in each column having indices between 1 and -4 (excluding -4) in the 2004 year DataFrame.
l = [(i,df_2004[df_2004[i] == -200].shape[0]) for i in df_2004.columns[1:-4]]
l

[('PT08.S1(CO)', 0),
 ('C6H6(GT)', 0),
 ('PT08.S2(NMHC)', 0),
 ('PT08.S3(NOx)', 0),
 ('PT08.S4(NO2)', 0),
 ('PT08.S5(O3)', 0),
 ('T', 0),
 ('RH', 0),
 ('AH', 0)]

In [ ]:
# S3.9: Again, calculate the percentage of rows containing '-200' in each column except for the 'DateTime', 'Year', 'Month' and 'Day' columns.
l = [(i,df_2005[df_2005[i] == -200].shape[0]) for i in df_2005.columns[1:-4]]
l

[('PT08.S1(CO)', 0),
 ('C6H6(GT)', 0),
 ('PT08.S2(NMHC)', 0),
 ('PT08.S3(NOx)', 0),
 ('PT08.S4(NO2)', 0),
 ('PT08.S5(O3)', 0),
 ('T', 0),
 ('RH', 0),
 ('AH', 0)]

Let's pause here. In the next class, we will learn how to group a DataFrame about a particular column to perform aggregation operations such as count, mean, median, and sum.

---